jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java

```diff
@@ -25,192 +25,256 @@
+    @Nullable private AuthenticatePushStep authenticatePushStep;
+    @Nullable private PullBaseImageStep pullBaseImageStep;
+    @Nullable private PullAndCacheBaseImageLayersStep pullAndCacheBaseImageLayersStep;
 
-  @Nullable
-  private ImmutableList<BuildAndCacheApplicationLayerStep> buildAndCacheApplicationLayerSteps;
+    @Nullable
+    private ImmutableList<BuildAndCacheApplicationLayerStep> buildAndCacheApplicationLayerSteps;
 
-  @Nullable private PushLayersStep pushBaseImageLayersStep;
-  @Nullable private PushLayersStep pushApplicationLayersStep;
-  @Nullable private BuildImageStep buildImageStep;
-  @Nullable private PushContainerConfigurationStep pushContainerConfigurationStep;
-  @Nullable private PushImageStep pushImageStep;
-  @Nullable private LoadDockerStep loadDockerStep;
-  @Nullable private WriteTarFileStep writeTarFileStep;
-
-  public StepsRunner(BuildConfiguration buildConfiguration) {
-    this.buildConfiguration = buildConfiguration;
+    @Nullable private PushLayersStep pushBaseImageLayersStep;
+    @Nullable private PushLayersStep pushApplicationLayersStep;
+    @Nullable private BuildImageStep buildImageStep;
+    @Nullable private PushContainerConfigurationStep pushContainerConfigurationStep;
+    @Nullable private PushImageStep pushImageStep;
+    @Nullable private LoadDockerStep loadDockerStep;
+    @Nullable private WriteTarFileStep writeTarFileStep;
+  }
 
+  /**
+   * Starts building the steps to run.
+   *
+   * @param buildConfiguration the {@link BuildConfiguration}
+   * @return a new {@link StepsRunner}
+   */
+  public static StepsRunner begin(BuildConfiguration buildConfiguration) {
     ExecutorService executorService =
         JibSystemProperties.isSerializedExecutionEnabled()
             ? MoreExecutors.newDirectExecutorService()
             : buildConfiguration.getExecutorService();
-    listeningExecutorService = MoreExecutors.listeningDecorator(executorService);
+
+    return new StepsRunner(MoreExecutors.listeningDecorator(executorService), buildConfiguration);
+  }
+
+  private final Steps steps = new Steps();
+
+  private final ListeningExecutorService listeningExecutorService;
+  private final BuildConfiguration buildConfiguration;
+
+  /** Collects the functions to start running the steps. */
+  private final List<Runnable> stepsRunnables = new ArrayList<>();
+
+  private StepsRunner(
+      ListeningExecutorService listeningExecutorService, BuildConfiguration buildConfiguration) {
+    this.listeningExecutorService = listeningExecutorService;
+    this.buildConfiguration = buildConfiguration;
   }
 
-  public StepsRunner runRetrieveTargetRegistryCredentialsStep() {
-    retrieveTargetRegistryCredentialsStep =
-        RetrieveRegistryCredentialsStep.forTargetImage(
-            listeningExecutorService, buildConfiguration);
+  public StepsRunner retrieveTargetRegistryCredentials() {
+    stepsRunnables.add(
```

Comment:
If all you need is a number of steps added, another way to achieve this is, instead of maintaining a list of runnables, you have a single runnable field, and do

```java
stepsAdded++;
Runnable previousRunnable = stepsRunnable;
stepsRunnable = () -> {
  previousRunnable.run();
  steps.retrieveTargetRegistryCredentialsStep = ...
}
```
And `runStepsRunners()` just becomes `stepsRunnable.run()`. But I think the list is OK too. I was just thinking maintaining a list might make readers think you really need to maintain a list of these for some reason.

Reply:
Sounds good! Will do.